### ISSUE TITLE
Action field in context is redundant

### ISSUE THREAD

**Issue by pkapustin** [beckn/protocol-specifications#180]
Context `action` field is always redundant in HTTP-based implementations. This will create problems if search endpoint is called and "confirm" is passed as action.
I suggest to remove it from the OpenAPI spec.

**Comment by core-wg-admin** [comment]
Beckn protocol is transport agnostic. So `action` cannot be removed from `context`.

**Comment by pkapustin** [comment]
This version of Beckn protocol uses OpenAPI, it is HTTP-specific by definition right?
Can we agree that this version of Beckn protocol is for HTTP?
If necessary, we can later come up with a different specification for other transport means, that doesn't use OpenAPI.

**Comment by core-wg-admin** [comment]
Open API has been used to document the APIs but the intent is to make the communication transport layer agnostic. At the time, we were not aware of Async API so we went with Open API but beckn was never designed to be exclusively for HTTP. It was more of an limitation in Open API than a design choice. Therefore, I suggest publishing the API document in Async API format and keep the action field rather than removing it and re-adding it later.

Secondly, the schema has to be made independent of the API and hence I suggest using the JSON Schema standard instead of Open API.

**Comment by techframewirk** [comment]
@core-wg-admin : 
> You can however remove it in network policy in consensus with other network participants.

Commenting on this point from the PR discussion. Would it be wise to remove it from a network level implementation while the property is still in the specification? There might be implementations that use the property in it's code even though its redundant when taking into account the http endpoint. These implementations will not work with the network in case the network chooses to remove the action property in context. This would decrease interoperability.

**Comment by pkapustin** [comment]
@core-wg-admin Okay sure, we can work towards that, but then let's make an HTTP-specific adaptation of the Beckn spec. There are many things to be shared there, for example related to HTTP error codes, so I think it would be definitely beneficial to have that standardized rather than each network coming up with their own conventions.

**Comment by pkapustin** [comment]
@techframewirk Network policy should map rather exactly to the implementations right? So if it is removed in the network policy, it should be removed from the implementations too.

**Comment by venkatramanm** [comment]
What happens if these messages are persisted for posterity for purpose of resolving disputes or for tracking, without action, we will not know what it was for. Whether it was for "init" or "confirm" for example.. We can put it externally but the signature would not  include the action. Today, we don't include the  url in the signing algorithm. Since the signature is only for the payload, I suggest, we retain the "action" attribute in the context.

**Comment by pkapustin** [comment]
@venkatramanm If we want to persist the messages, we can include action separately right?
Also, I think url should be included in the signing algorithm, wouldn't it be better? The problem with having action in the context is that there is always a chance someone passes wrong action, as long as there is a specific endpoint you are calling. So the endpoint is always the source of truth, not the action in the context.

**Comment by techframewirk** [comment]
@pkapustin I understand network policy should map towards implementations. But `action` is still part of the core spec. It is reasonable to have code that uses `context.action` assuming it will be there. So I think we should not violate protocol until it becomes a standard. 

I want to make this point not just for this specific field we are discussing, but in general. Say there are multiple different networks. It would be quite counter productive if all of these networks deviate from the core spec/domain spec in their own ways. The changes required for a technical solution in the same domain from one network to plug in to another network should be as minimal as possible. I feel the more that increases the less the utility of the protocol. Which is a reason why I think we should be very careful while violating the specs in a network. Let me know what you think!

Having wrong action compared to the endpoint **should** return NACK.

**Comment by core-wg-admin** [comment]
The protocol should clearly specify that on HTTP implementations, `context.action` value should be equal to the URL endpoint. That should close this issue.

**Comment by pkapustin** [comment]
@techframewirk 

> I want to make this point not just for this specific field we are discussing, but in general. Say there are multiple different networks. It would be quite counter productive if all of these networks deviate from the core spec/domain spec in their own ways. The changes required for a technical solution in the same domain from one network to plug in to another network should be as minimal as possible. I feel the more that increases the less the utility of the protocol. Which is a reason why I think we should be very careful while violating the specs in a network. Let me know what you think!

I agree, this is one reason for domain specs being so important, they will allow us to minimize the differences between the network specs.

@techframewirk @core-wg-admin 

> @pkapustin I understand network policy should map towards implementations. But action is still part of the core spec. It is reasonable to have code that uses context.action assuming it will be there. So I think we should not violate protocol until it becomes a standard.

> The protocol should clearly specify that on HTTP implementations, context.action value should be equal to the URL endpoint. That should close this issue.

I think this is an example of an issue that is easily solved by creating an HTTP adaptation of the spec: we can keep action in the core spec and remove it in the HTTP adaptation, avoiding all the problems with redundancy in HTTP implementations.
See also https://github.com/beckn/protocol-specifications/issues/186.

**Comment by venkatramanm** [comment]
WE today sign only the payload and NOT the url. Many apis like init and confirm could look very similar. If tomorrow to resolve disputes, if we need to prove using signed payload that a confirm or on_confirm were indeed called. we will not be able to say for sure even if we persist the messages as some one can easily tell that it is a init json and not confirm json. But "if" the action was in the context, it is included in the contract being signed. Hence we should not remove the action from the context! in any spec.. unless the url is included in the signing algorithm. For now i think we should not  remove the action.

**Comment by pkapustin** [comment]
I agree with you that we should include the url in the signing algorithm. This should be done in any case, because otherwise it may be a security problem, right? Imagine a message is intercepted and sent to a different endpoint. I think it is important that this is detected during signature verification. Relying on the endpoint handler later checking the action is not good enough especially as these checks will often be missing (as long as the field is otherwise redundant).
So, I think this is solved by including url in the signature.
